### PR TITLE
Add Clang Vector Extension and swich to xyzw quat

### PR DESCRIPTION
**Pull request by jopadan**
- use ext_vector_type for -DCMAKE_C_COMPILER=clang
- pass vector argument as static instance of type not address (they are all const anyway)
- use xyzw components instead of wxyz for quat
- add [src/]CMakeFiles to .gitignore

**TODO**
- Assert Gyro Gamepad functioning as before (I was not able to get a Gyro capable device)

**DONE**
- build/run with gcc/clang
- assert non-gyro gamepad function

### CODE REVIEW

File: cmake/WoofSettings.cmake [hunk 1]
```diff
@@ -74,6 +75,7 @@ if(MSVC)
 else()
     # We only want -Wall on GCC compilers, since /Wall on MSVC is noisy.
     _checked_add_compile_option(-Wall)
+    _checked_add_compile_option(-std=gnu23)
```
Comment (fabiangreffrath) [inline cmake/WoofSettings.cmake:78]: Why is this required? We use union-of-function-pointers type quirks in the code which are valid in C99 but emit warnings with C23. The exact warnings that you override in your change in few lines above. Getting rid of these warnings would require a major work on the code that initializes the states and cheats arrays, or sticking to C99 for the time being.

**Review by fabiangreffrath — COMMENTED** [review]
Thank you very much for your contribution! While it looks very interesting and promising, I am not sure which changes are really required and which may just be considered drive-by (i.e. the change to C23).

File: src/m_vector.h [hunk 2]
```diff
@@ -22,178 +22,186 @@
 
 #include "doomtype.h"
 
+#ifdef __clang__
```
Comment (fabiangreffrath) [inline src/m_vector.h:25]: "Query for this feature with `__has_attribute(ext_vector_type)`"

https://clang.llvm.org/docs/LanguageExtensions.html

File: cmake/WoofSettings.cmake (hunk 1 above)
Comment (jopadan) [reply cmake/WoofSettings.cmake:78]: usleep and fileno only work with -std=gnu23 using clang and gcc

Comment (rfomin) [reply cmake/WoofSettings.cmake:78]: > usleep and fileno only work with -std=gnu23 using clang and gcc

Why? We use usleep in `I_SleepUS` function, seems to work with C99: https://github.com/fabiangreffrath/woof/blob/8a736339a9bfd25032498147a55c656d03519a3f/src/i_timer.c#L226

File: src/doomtype.h [hunk 3]
```diff
@@ -90,6 +91,18 @@ typedef byte lighttable_t;
 #define inline __inline
 #endif
 
+#if defined(__clang__)
```
Comment (fabiangreffrath) [inline src/doomtype.h:94]: Could you indent this as seen in the other defines?

File: src/m_vector.h [hunk 4]
```diff
@@ -22,178 +22,186 @@
 
 #include "doomtype.h"
 
+#if defined(__clang__) && __has_attribute(ext_vector_type)
```
Comment (fabiangreffrath) [inline src/m_vector.h:25]: Could you abstract this away and use the corresponding vector extension on gcc?

https://gcc.gnu.org/onlinedocs/gcc/Vector-Extensions.html

File: src/doomtype.h (hunk 3 above)
Comment (jopadan) [reply src/doomtype.h:94]: No because gcc vector extension doesn't support xyzw components.
The only way to support it is to switch to array access operator[] for the existing struct vector types.

File: src/m_vector.h (hunk 4 above)
Comment (jopadan) [reply src/m_vector.h:25]: No because gcc vector extension doesn't support xyzw components.
The only way to support it is to switch to array access operator[] for the existing struct vector types.

Comment (fabiangreffrath) [reply src/m_vector.h:25]: I see, thank you for elaborating. @ceski-1 would you be so kind and check if this breaks anything on your end? I still don't have a gamepad with gyro and rumble, but Christmas is around the corner. :wink:

Comment (ceski-1) [reply src/m_vector.h:25]: Gyro works fine for me.

File: src/doomtype.h [hunk 5]
```diff
@@ -23,6 +23,7 @@
 
 #include <stddef.h> // size_t, NULL
 #include <stdint.h> // [FG] intptr_t types
+#include <stdbool.h>
```
Comment (rfomin) [inline src/doomtype.h:26]: Please remove `#include <stdbool.h>` from `doomtype.h`, we don't use `bool`. There is a `boolean` type in the Doom codebase that has a different size.

Comment (rfomin) [inline src/doomtype.h:26]: Please remove `#include <stdbool.h>` from `doomtype.h`, we don't use `bool`. There is a `boolean` type in the Doom codebase that has a different size.

File: .gitignore [hunk 6]
```diff
@@ -14,3 +14,5 @@ CMakeSettings.json
 
 # clangd
 /.cache/
+CMakeFiles/
+src/CMakeFiles/
```
Comment (rfomin) [inline .gitignore:18]: Why this change?

File: src/doomtype.h [hunk 7]
```diff
@@ -90,6 +91,18 @@ typedef byte lighttable_t;
 #define inline __inline
 #endif
 
+#if defined(__clang__)
+ #ifndef __has_feature
+  #define __has_feature(x) 0
```
Comment (fabiangreffrath) [inline src/doomtype.h:96]: Why do we even need this? This is a clang-only feature and the clang docs are pretty clear about the check.

**Comment by gendlin** [comment]
Would now be a good time to start compiling with `-ffastmath`?

File: .gitignore (hunk 6 above)
Comment (jopadan) [reply .gitignore:18]: To avoid in source tree build generated files being listed in git add -A and git commit -a

File: src/doomtype.h (hunk 7 above)
Comment (jopadan) [reply src/doomtype.h]: I'v removed it now.

**Comment by fabiangreffrath** [comment]
I think you should move the feature check to the build system like this:

https://github.com/fabiangreffrath/woof/blob/e8321801319195bd96ce4e0a38e9b1a1e0fcddfb/CMakeLists.txt#L68-L77

**Comment by rfomin** [comment]
> Would now be a good time to start compiling with `-ffast-math`?

I think we still have too few float calculations.

File: .gitignore (hunk 6 above)
Comment (rfomin) [reply .gitignore:18]: If you create a folder named "build*", they should already be ignored. For example:
```
cd woof
cmake -B build -G Ninja
```

Comment (fabiangreffrath) [reply .gitignore:18]: Indeed, this change is unrelated.

**Comment by fabiangreffrath** [comment]
> I think you should move the feature check to the build system like this:

Could you please apply something like this (untested) and see if it works for you?

```patch
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 48b1493b..5185f3cd 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -75,6 +75,16 @@ check_c_source_compiles("
   "
   HAVE__DIV64
 )
+check_c_source_compiles("
+    int main()
+    {
+        typedef float vec __attribute__((ext_vector_type(4)));
+        vec a = (vec){0, 1, 2, 3};
+        return 0;
+    }
+  "
+  HAVE_EXT_VECTOR_TYPE
+)

 option(CMAKE_FIND_PACKAGE_PREFER_CONFIG
        "Lookup package config files before using find modules" ON)
diff --git a/src/m_vector.h b/src/m_vector.h
index 35373549..74755f8b 100644
--- a/src/m_vector.h
+++ b/src/m_vector.h
@@ -20,9 +20,10 @@

 #include <math.h>

+#include "config.h"
 #include "doomtype.h"

-#if defined(__has_attribute) && defined(__clang__) && __has_attribute(ext_vector_type)
+#if defined(HAVE_EXT_VECTOR_TYPE)
 typedef float vec __attribute__((ext_vector_type(3)));
 #else
 typedef struct
@@ -33,7 +34,7 @@ typedef struct
 } vec;
 #endif

-#if defined(__has_attribute) && defined(__clang__) && __has_attribute(ext_vector_type)
+#if defined(HAVE_EXT_VECTOR_TYPE)
 typedef float quat __attribute__((ext_vector_type(4)));
 #else
 typedef struct
```

**Review by fabiangreffrath — APPROVED** [review]
LGTM, I don't care that much about the .gitignore changes.

**Review by ceski-1 — APPROVED** [review]
LGTM

**Comment by rfomin** [comment]
Thanks!

File: src/m_vector.h [hunk 8]
```diff
@@ -20,180 +20,189 @@
 
 #include <math.h>
 
+#include "config.h"
 #include "doomtype.h"
 
+#if defined(HAVE_EXT_VECTOR_TYPE)
```
Comment (fabiangreffrath) [inline src/m_vector.h:26]: I think we never actually use these, because we don't have `#cmakedefine HAVE_EXT_VECTOR_TYPE` in our `config.h.in`, right?
